File: AutoGis/AutoGis/AutoGis/backend/internal/usecase/user_access.go
package usecase

import (
	"time"

	"github.com/gmt061/autogis-backend/internal/domain"
)

func deriveUserAccountType(role domain.UserRole) domain.UserAccountType {
	switch role {
	case domain.RoleMaster:
		return domain.UserAccountTypePrivateExecutor
	case domain.RoleAutoWash:
		return domain.UserAccountTypeAutoWash
	case domain.RoleAutoShop:
		return domain.UserAccountTypeAutoShop
	case domain.RoleAutoService:
		return domain.UserAccountTypeAutoService
	case domain.RoleAdmin:
		return domain.UserAccountTypeAdmin
	case domain.RoleModerator:
		return domain.UserAccountTypeModerator
	default:
		return domain.UserAccountTypeCustomer
	}
}

func hasBusinessCrmAccess(_ *domain.User, accountType domain.UserAccountType) bool {
	switch accountType {
	case domain.UserAccountTypeAutoWash,
		domain.UserAccountTypeAutoShop,
		domain.UserAccountTypeAutoService:
		// Бизнес-CRM будет включаться отдельно, после появления подписки.
		return false
	default:
		return false
	}
}

func deriveUserCapabilities(user *domain.User, accountType domain.UserAccountType) domain.UserCapabilities {
	professionalCabinet := user != nil && user.IsProfessional
	crmMini := professionalCabinet && accountType == domain.UserAccountTypePrivateExecutor
	businessCrm := professionalCabinet && hasBusinessCrmAccess(user, accountType)
	providerWorkflowAccess := professionalCabinet

	return domain.UserCapabilities{
		ProfessionalCabinet: professionalCabinet,
		CrmMini:             crmMini,
		BusinessCrm:         businessCrm,
		ProfessionalChat:    providerWorkflowAccess,
		Applications:        providerWorkflowAccess,
		Calendar:            providerWorkflowAccess,
	}
}

func buildUserResponse(user *domain.User) *domain.UserResponse {
	if user == nil {
		return nil
	}

	privacyAgreedAt := ""
	if user.PrivacyAgreedAt != nil {
		privacyAgreedAt = user.PrivacyAgreedAt.Format(time.RFC3339)
	}

	accountType := deriveUserAccountType(user.Role)

	return &domain.UserResponse{
		ID:              user.ID,
		Phone:           user.Phone,
		Name:            user.Name,
		ContactNumber:   user.ContactNumber,
		Role:            user.Role,
		AccountType:     accountType,
		IsProfessional:  user.IsProfessional,
		Capabilities:    deriveUserCapabilities(user, accountType),
		AgreedToPrivacy: user.AgreedToPrivacy,
		PrivacyAgreedAt: func() *string {
			if privacyAgreedAt == "" {
				return nil
			}
			return &privacyAgreedAt
		}(),
		Coordinates: user.Coordinates,
		CreatedAt:   user.CreatedAt.Format(time.RFC3339),
		UpdatedAt:   user.UpdatedAt.Format(time.RFC3339),
	}
}

File: AutoGis/AutoGis/src/common/lib/userAccess.ts
import {
    User,
    UserAccountType,
    UserCapabilities,
    UserRole,
} from "@common/types/user";

export type UserCapabilityKey = keyof UserCapabilities;

export function deriveUserAccountType(role?: UserRole): UserAccountType {
    switch (role) {
        case UserRole.MASTER:
            return UserAccountType.PRIVATE_EXECUTOR;
        case UserRole.AUTO_WASH:
            return UserAccountType.AUTO_WASH;
        case UserRole.AUTO_SHOP:
            return UserAccountType.AUTO_SHOP;
        case UserRole.AUTO_SERVICE:
            return UserAccountType.AUTO_SERVICE;
        case UserRole.ADMIN:
            return UserAccountType.ADMIN;
        case UserRole.MODERATOR:
            return UserAccountType.MODERATOR;
        default:
            return UserAccountType.CUSTOMER;
    }
}

export function deriveUserCapabilities(input: {
    accountType?: UserAccountType;
    role?: UserRole;
    isProfessional?: boolean;
}): UserCapabilities {
    const accountType =
        input.accountType ?? deriveUserAccountType(input.role);
    const professionalCabinet = input.isProfessional === true;
    const crmMini =
        professionalCabinet &&
        accountType === UserAccountType.PRIVATE_EXECUTOR;
    const businessCrm = false;
    const providerWorkflowAccess = professionalCabinet;

    return {
        professionalCabinet,
        crmMini,
        businessCrm,
        professionalChat: providerWorkflowAccess,
        applications: providerWorkflowAccess,
        calendar: providerWorkflowAccess,
    };
}

export function normalizeUserProfile(profile: User): User {
    const accountType =
        profile.accountType ?? deriveUserAccountType(profile.role);
    const hasProfessionalActivity =
        Array.isArray(profile.activityTypes) &&
        profile.activityTypes.length > 0;
    const isProfessional = profile.isProfessional ?? hasProfessionalActivity;

    return {
        ...profile,
        accountType,
        contactNumber: profile.contactNumber ?? profile.phone,
        isProfessional,
        capabilities:
            profile.capabilities ??
            deriveUserCapabilities({
                accountType,
                role: profile.role,
                isProfessional,
            }),
    };
}

export function hasCapability(
    profile: User | null | undefined,
    capability: UserCapabilityKey,
): boolean {
    if (!profile) return false;
    const normalized = normalizeUserProfile(profile);
    return normalized.capabilities?.[capability] === true;
}
